### DISCUSSION TITLE
Class Middleware with Parameters and DI

### DISCUSSION THREAD

**Discussion by tzarger** [MichalLytek/type-graphql#1232] [Q&A] [answered]
Sorry for all the posts lately... but I am trying to leverage lots of the greatness in the project.

I am using TSyringe for DI and trying to use Middleware to roll my own auth checker, since I have different needs for different resolvers and can only have one instance of an auth checker.

The issue, I found another post about wrapping the class with a function that allows input paramters:

`@UseMiddleware(myFunction('Hello', 'World'))`

This unfortunately does not work since I am not able to register a function, only classes in TSyringe. How would you pass parameters into a class based middleware that can also work with Di?

Thanks so much for you thoughts.

**Comment by MichalLytek** [comment, marked as answer]
>  I am not able to register a function, only classes in TSyringe.

You would get only single instance from container, while you need multiple because of the custom parameters you provide (`'Hello', 'World'`).

**Reply by tzarger** [reply]
@MichalLytek with your response, I am guessing there is no possibility of a Class Based Middleware with parameters so that it can be used with TSyringe? I assume that goes for any kind of decorator of any kind?

In order to use many of these features, it seems that I would have to forego using the container goodness that this project offers with the wondeful baked in constructor injection...

**Reply by MichalLytek** [reply]
`context.container.resolve(YourClass)` is your solution

**Reply by tzarger** [reply]
I found out how to write the Middlware as a class and get it registered.

```
import { MiddlewareInterface, NextFn, ResolverData } from 'type-graphql';

import { GraphQLContext } from '@/server/context';
import { Logger } from '@/services/Logger';
import { injectable } from 'tsyringe';

export function AuthGuard(role: string, permission: string) {
  @injectable()
  class AuthGuardImplementation implements MiddlewareInterface<GraphQLContext> {
    constructor(private readonly logger: Logger) {}

    async use({ root, args, context, info }: ResolverData<GraphQLContext>, next: NextFn, test?: string) {
      this.logger.log(`Logging access: ${context.user?.id} -> ${info.parentType.name}.${info.fieldName}`, role, permission);
      return next();
    }
  }

  return AuthGuardImplementation;
}
```

Intead of returning the class inline, creating the class then returning allowed me to put the @injectable() on the class and it works now with constructor injection as well.
